### PR TITLE
Move cluster builder, sdfgi and gi structures to clustered renderer, and more

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Work in progress...

Last big part of the render reorg we're doing, this PR once finished will target the following parts:
- Move SSEffects from `RenderSceneRenderRD` into `RenderForwardClustered` **(done)**
- Move cluster builder objects from `RenderSceneRenderRD` into `RenderForwardClustered` **(done)**
- Move `LightInstance`, `ReflectionAtlas`, `LightmapInstance` and `ShadowAtlas` from `RenderSceneRenderRD` into `LightStorage` **(done)**
- Move `DecalInstance` from `RenderSceneRenderRD` into `TextureStorage` **(done)**
- Reorganise `RenderSceneRenderRD::render_scene` so it sets up our scene data and render data structures and then calls `_render_scene`, move further setup code into `_render_scene` as most logic was only applicable to clustered renderer **(done)**
- Move `RenderSceneRenderRD::_pre_opaque_render` into our implementation classes. Where code is shared, moved that into new methods in `RenderSceneRenderRD` **(done)**
- Move shadow logic into our implementation classes as these likely will become very different in the mobile renderer **(done)**
- Finish `ForwardID` changes **(done)**

A few notes about moving code from `RenderSceneRenderRD` into the clustered and mobile renderers:
- There is still a fair amount of code that uses pointers to structures that can't be exposed to GDExtensions but structural this now seems sound and is more a matter of adding getters and setters and turning some of these structures into proper classes (mainly `RenderDataRD`).
- Some code is being duplicated but the expectation here is that this code in the mobile renderer will get further rewrites as we optimise things for mobile GPUs and this code will thus start deviating greatly.
- Some code that has been moved into the clustered renderer might need to be moved into the effects classes, functions like `_process_ssr` and `_process_ssao` and such, though the bulk of the logic is already in the effects classes.

To be further discussed (possibly to do in later PRs as these are not critical for 4.0):
- Make `RenderDataRD` subclassed from RefCounted just like we did with SceneDataRD and RenderBuffersRD
- Start looking into what getters/setters need to be added to make extensions possible.
- Move clustered renderer and mobile renderer implementations out of servers and into modules.

### CODE REVIEW

File: servers/rendering/renderer_rd/shaders/forward_clustered/SCsub
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/shaders/forward_clustered/SCsub:7]: @clayjohn this should ensure that shaders are re-compiled when includes in parent folder change.

Should we put this in a separate PR or just wait until this is merged?

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/renderer_scene_render_rd.h:249]: I moved these here so they are accessible by our subclasses as I moved methods that only belong in our clustered forward renderer into that class. But this really needs to change to getters.
I also believe these can be moved into our effects class and in doing so simplifying the API of our effects class.

File: servers/rendering/renderer_rd/storage_rd/forward_id.h
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/storage_rd/forward_id.h:40]: Due to this now being accessed from various places it seemed better to move these into their own header files.

I'm thinking about moving the related methods into a support class as well which we can subclass for the mobile renderer functionality.

File: servers/rendering/renderer_rd/storage_rd/light_storage.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/storage_rd/light_storage.h:134]: This is now all cleaned up, just `DecalInstance` is left.

**Comment by BastiaanOlij** [comment]
I've changed this to ready for review, there are a few small things I still want to do but all the big structural things are done and working.

File: servers/rendering/renderer_rd/shaders/forward_clustered/SCsub
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/shaders/forward_clustered/SCsub:7]: This is now in https://github.com/godotengine/godot/pull/66423

**Comment by BastiaanOlij** [comment]
I think I've taken this as far as I can. There is probably some more polish worth doing but I think this needs to be discussed in a wider group.

@clayjohn @reduz @JFonS we should probably have another render meeting. It would be good if you all can review this ASAP and look into whether what we have can be merged seeing this work will clash with work of others.

We can then dot any i's and cross any t's we come up with in a follow up PR but I think all the important stuff is in here.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (reduz) [inline servers/rendering/renderer_rd/renderer_scene_render_rd.h:101]: IMO all this should be moved to LightStorage (decal to TextureStorage?) there is not much of a reason to unify thesesince its just an array that you need to pass to the render shader, which can simply be obtained with a function and which will not change from frame to frame.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (reduz) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.h:248]: Yeah it looks to me that these should be moved to their own class.

**Comment by reduz** [comment]
IMO the biggest thing that should be missing is that the cluster (which probably should be renamed to something else) information should likely be in the same place as the instances, since this is something static that needs to be filled basted on a list of instances.

**Comment by reduz** [comment]
Likewise, the shadow atlas information should also be in the same place as the light instances.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.h:101]: Agreed, I shall work on that. As I mentioned in chat, should we do this as a separate PR and merge the work already done? This PR is getting fairly large.

**Comment by BastiaanOlij** [comment]
Ok, I've moved the code for DecalData (Decal UBO) into texture storage, I'll do the same with the other parts of `SceneLightDataRD` but moving them into light storage. Was easier then I thought.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.h]: This is now completed, `SceneLightDataRD` is no more! All data structures and code are now in `LightStorage` and `TextureStorage`.

These buffers are depth sorted and distance fade is applied so I do think it currently updates frame by frame but all in all, it is all a lot cleaner now.

There are still a few ties to the render implementation that could be improved but I think those are not critical.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.h:116]: I need to look into this further because some of these are a little entangled. I'd like to do these in future PRs.

**Comment by BastiaanOlij** [comment]
@reduz, @clayjohn I'm wondering if we should move and duplicate `RendererSceneRenderRD::_render_shadow_pass` into the two render implementations and then remove `_render_shadow_begin`, `_render_shadow_append`, `_render_shadow_process` and `_render_shadow_end` as virtual functions from `RendererSceneRenderRD`.
I suspect that in the mobile renderer this logic will be rewritten and possibly solved very differently and these will be methods that will only exist in the clustered renderer.

Worth doing in a separate PR though I think.

**Comment by BastiaanOlij** [comment]
> Likewise, the shadow atlas information should also be in the same place as the light instances.

This is now done.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.h]: This is done.

**Comment by BastiaanOlij** [comment]
2nd commit moves settings into our SSEffects class, still hunting a bug that seems to have broken SSIL

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.h]: Managed to move all the SSEffects related settings into the SSEffects class. This is in a separate commit as it really stands apart from the main body of work.

There is more to move but as I think those deserve smaller PRs. This was the bigger one.

**Comment by BastiaanOlij** [comment]
@clayjohn @akien-mga ok I thought something had broken in effects with the move of the settings but things don't seem to be in order in the current master. I'll have to look at that separately but it's not due to this PR.

Clayjohn, if you could give it a quick look, I realise it all being in one big commit makes it harder to see the changes reduz asked for, but I think we should move to merging this if CI succeeds.

File: servers/rendering/renderer_rd/renderer_scene_render_rd.cpp
Comment (clayjohn) [inline servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:3817]: You missed moving ``environment_set_ssao_quality()`` its just 6 lines up, but github won't let me comment on the line directly

File: servers/rendering/renderer_rd/storage_rd/utilities.cpp
Comment (clayjohn) [inline servers/rendering/renderer_rd/storage_rd/utilities.cpp:110]: We shouldn't have the ``else`` condition here. This makes it so the function always returns true even if it didn't free anything

**Review by clayjohn — COMMENTED** [review]
Amazing work! Just two little fixes and then I think we are ready to merge this!!!!

File: servers/rendering/renderer_rd/renderer_scene_render_rd.cpp
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:3817]: Also missed `environment_set_ssil_quality` :)

Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:3817]: I'm wondering if we should write them out?

File: servers/rendering/renderer_rd/storage_rd/utilities.cpp
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/storage_rd/utilities.cpp:110]: I think that`s just the diff making it look weird, the code is like this:
```
	} else if (RendererRD::Fog::get_singleton()->owns_fog_volume(p_rid)) {
		RendererRD::Fog::get_singleton()->fog_volume_free(p_rid);
		return true;
	} else if (owns_visibility_notifier(p_rid)) {
		visibility_notifier_free(p_rid);
		return true;
	}

	return false;
}
```

So there is no else here, it hits the `return false;` at the end if nothing is freed. 

That said, I'm not against turning it into:
```
	} else if (RendererRD::Fog::get_singleton()->owns_fog_volume(p_rid)) {
		RendererRD::Fog::get_singleton()->fog_volume_free(p_rid);
		return true;
	} else if (owns_visibility_notifier(p_rid)) {
		visibility_notifier_free(p_rid);
		return true;
	} else {
		return false;
	}
}
```
So that it is absolutely clear it's supposed to exit with false if nothing is freed..

File: servers/rendering/renderer_rd/renderer_scene_render_rd.cpp
Comment (clayjohn) [reply servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:3818]: I don't have a preference either way

**Review by clayjohn — APPROVED** [review]
Great work! Looks fine to me. Let's double check with @reduz If he wants to take another look before this is merged, but I am happy to merge as-is

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
